### PR TITLE
feat(Bank Transaction): auto-book included bank fees

### PR DESCRIPTION
**Pull request by barredterra**
## Summary

Auto-book included bank fees on Bank Transaction submit (withdrawals) and at reconciliation time (deposits).

**Withdrawals**: On submit, a Journal Entry is created (`Dr Bank Fees / Cr Bank`) and allocated against the Bank Transaction, partially reconciling the fee portion.

**Deposits**: Fee booking is deferred to reconciliation, where the correct counter-account (e.g. receivable) is known. The reconciliation tool matches deposits using `deposit + included_fee` as the target amount, so the right vouchers are found. At reconciliation:
- **Single-currency**: A Journal Entry settles both the fee and the invoice in one entry (`Dr Bank, Dr Bank Fees / Cr Receivable`)
- **Multi-currency**: A Payment Entry is created with a deduction row for the fee. The system exchange rate determines exchange gain/loss separately from the fee.

### Configuration

- New `Bank Fee Account` field on Bank Account (Link to Expense account, same currency)
- New `Enable Automatic Journal Entries for Bank Fees` checkbox in Banking Settings
- All company Bank Accounts must have a fee account before the feature can be enabled

### Depends on

- frappe/erpnext#52760 (min ERPNext version bump)
- ~frappe/erpnext#54269 (`unallocated_amount` for `deposit=0 + included_fee` transactions)~ case excluded via e03d141

## Test plan

- [ ] Submit withdrawal with included fee → fee JE created and allocated
- [ ] Submit deposit with included fee → no fee JE created
- [ ] Reconcile single-currency deposit with fee against Sales Invoice → JE with bank + fee rows, invoice fully cleared
- [ ] Reconcile multi-currency deposit with fee against foreign invoice → PE with fee deduction, exchange gain/loss handled separately
- [ ] Feature disabled → no fee JE on submit, no fee handling at reconciliation
- [ ] Bank Account without fee account rejects save when feature is enabled
- [ ] Currency mismatch between bank account and fee account is rejected
- [ ] Cancel Bank Transaction → linked fee JEs are cancelled
- [ ] Fee larger than withdrawal is rejected

### CODE REVIEW

**Comment by barredterra** [comment]
Proposed fixes:

## 1) Cancel only fee JEs created by this feature ✅ 

Depends on https://github.com/frappe/erpnext/pull/52760

Use **first-class JE row references** to mark ownership, not `cheque_no` heuristics.

- **Ownership model**
  - Treat fee JEs as linked vouchers by setting `Journal Entry Account.reference_type = "Bank Transaction"` and `reference_name = <bt.name>`.
  - Also set `Journal Entry.is_system_generated = 1` when creating these fee JEs.
  - Keep `cheque_no` if useful for reporting/search, but do not use it as the primary cancel selector.
  - Cancellation of linked JE is handled by framework

- **Tests to add**
  - `create_automatic_journal_entry` writes JE row reference + `is_system_generated`.
  - Cancel BT cancels only linked system-generated fee JEs.
  - Unrelated JE with same `cheque_no` is untouched.

---

## 2) Make guard clauses feature-aware and migration-friendly ✅ 

**Problem**  
`before_submit` raises on missing `bank_account` even when fee automation is disabled. That can break legacy flows unnecessarily.

**Fix**  
Reorder guards in `before_submit()`:

1. read flag `enable_automatic_journal_entries_for_bank_fees`
2. if flag is off -> return immediately
3. if `included_fee <= 0` -> return immediately
4. only then validate prerequisites (`bank_account`, fee constraints, etc.) and try creating fee JE

**Compatibility choice**
- For missing setup (`bank_account` / `bank_fee_account`):
  - **strict mode**: `frappe.throw` (safer accounting)
  - **compat mode**: `frappe.log_warning` + return (safer migrations)
- Given your migration concern, compat mode is reasonable initially.

**Test to add**
- With feature flag OFF and missing `bank_account`, submission path does not fail due to this hook.
- With feature flag ON and valid fee case, hook runs as expected.

---

## 3) Stop overwriting allocation totals manually ✅ 

**Problem**  
`create_je_bank_fees` manually sets `allocated_amount`/`unallocated_amount`, which can conflict with existing rows or previous allocations.

**Fix**  
Let core calculation do the math.

- Keep appending the JE payment row (`allocated_amount = included_fee` for withdrawal, `0` for deposit)
- Then call:
  - `doc.update_allocated_amount()`
- Then set status based on resulting unallocated value:
  - `Reconciled` if `<= 0`
  - else `Unreconciled`

**Why this is best**  
Avoids brittle arithmetic and stays aligned with ERPNext’s own allocation behavior.

**Test to add**
- Withdrawal with pre-existing payment entries: total allocation remains correct after fee JE append.
- Edge case: `withdrawal == included_fee` ends as fully reconciled.

---

## 4) Avoid partial cancel by failing if fee JE cancellation fails ✅ 

**Problem**  
`on_cancel` currently swallows JE cancel exceptions via `msgprint`, which can leave inconsistent state.

**Fix**
- Collect cancellation failures
- If any fail, throw a single `ValidationError` with JE names/links
- This aborts BT cancel and keeps consistency

**Why this is best**  
Prevents “BT canceled but linked fee JE still submitted” inconsistencies.

**Test to add**
- Mock JE `.cancel()` to raise
- Assert BT cancellation is blocked and error message includes the failed JE.

**Comment by 0xD0M1M0** [comment]
"Unreconcile Transaction" in the _Bank Transaction_ should not unlink the fee entry.

**Comment by barredterra** [comment]
@0xD0M1M0  I’d like your thoughts on whether it conceptually makes sense to create bank-fee **Journal Entry** records for deposit **Bank Transactions** at submit time.

For withdrawals, this seems reasonable. But for deposits, I’m not sure it is. My concern is that the correct counter-account is usually not known yet when the **Bank Transaction** is submitted. We may know the net amount that reached the bank and the `included_fee`, but we often do not yet know which receivable, advance, or other account should be cleared.

Example: we have a EUR bank account, a customer owes 100 USD, the exchange rate is 0.8, so the receivable is 80 EUR. The bank keeps 5 EUR in conversion fees and deposits 75 EUR. Economically, I would expect:

- Dr Bank 75 EUR
 - Dr Bank Fees 5 EUR
- Cr Accounts Receivable 80 EUR

The issue is that at **Bank Transaction** submit time, we usually do not yet know which receivable account should be credited. That makes the deposit fee booking feel incomplete or potentially misleading if we create a fee JE immediately.

So my question is: should deposit-side `included_fee` remain just statement data until reconciliation, when we know the real counter-account, or should we still create an immediate accounting entry, perhaps against some interim clearing account? Also, should we treat deposit fees differently from withdrawal fees for that reason?

**Comment by barredterra** [comment]
~Special case: 0 deposit and included fee -> can be booked directly.~ No -> can be used to offset receivables.
Create Journal Entry instead of Payment Entry.
When matching deposits in bank reco, add fee to matching amount.

**Comment by barredterra** [comment]
@greptileai

**Comment by 0xD0M1M0** [comment]
### Test plan

- [X] Submit withdrawal with included fee → fee JE created and allocated
- [X] Submit deposit with included fee → no fee JE created
- [X] Reconcile single-currency deposit (100€) with fee (13€) against Sales Invoice (120€) → JE with bank + fee rows, invoice partically cleared (7€)
- [X] Reconcile single-currency deposit (100€) with fee (13€) against Sales Invoice (113€) → JE with bank + fee rows, invoice fully cleared
- [X] BT1: Reconcile single-currency deposit (110€) with fee (13€) against Sales Invoice Si 1 (120€) -> Remaining 3€
- [X] BT1: Reconcile single-currency deposit (110€) with fee (13€) with 3€ unallocated against Sales Invoice SI 2 (120€) **-> Failed: Fee deducted again! Sales Invoice remaining 104€** -> 2262e12
- [x] Reconcile multi-currency deposit with fee against foreign invoice → PE with fee deduction, exchange gain/loss handled separately **-> Failed: Fee not deducted nor visible anywhere** -> difficult to fix because PE does not accept foreign-currency deductions -> ba60110
- [X] Feature disabled → no fee JE on submit, no fee handling at reconciliation
- [X] Bank Account without fee account rejects save when feature is enabled
- [X] +1 Enabling the feature is only possible if all bank accounts have a fee account
- [X] Currency mismatch between bank account and fee account is rejected
- [X] Cancel Bank Transaction → linked fee JEs are cancelled
- [X] Included fee larger than withdrawal is rejected

To be handled in a separate PR:

- [F] Unreconcile bank transaction should not remove the fee JE
- [F] Cancel Bank Transaction → linked JEs are cancelled **-> Failed: not canceled (exable BT1 vs. sales invoice SI 1 & SI 2**

**Comment by barredterra** [comment]
Simplify: disallow edge cases

- [x] 0 deposit with included fees
- [x] multiple partial reconciliations for BT with fees

Unreconcile bank transaction:

- should not unreconcile vouchers that were created for this BT to avoid double booking. Instead, flag newly created and cancel? -> separate PR

File: banking/klarna_kosma_integration/doctype/bank_reconciliation_tool_beta/unpaid_vouchers.py [hunk 1]
```diff
@@ -618,3 +679,55 @@ def get_outstanding_amount(payment_doctype, payment_name) -> float:
...
+	Returns 0.0 when the legacy pre-feature behavior should be preserved.
+	"""
+	if not (flt(bt.deposit) > 0 and flt(bt.included_fee) > 0):
+		return 0.0
+
+	if not flt(
+		frappe.db.get_single_value("Banking Settings", "enable_automatic_journal_entries_for_bank_fees")
+	):
+		return 0.0
+
```
[the remark this answers is not in the document]
Comment (barredterra) [reply banking/klarna_kosma_integration/doctype/bank_reconciliation_tool_beta/unpaid_vouchers.py:812]: Should be resolved via 2262e12

**Comment by barredterra** [comment]
@Mergifyio backport version-16-hotfix
